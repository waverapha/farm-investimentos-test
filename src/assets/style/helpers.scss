@import './_variables';

.container {
  max-width: 80rem;
  margin: 0 auto;
}

.flex-container {
  display: flex;
  padding: 0 1rem;
  @extend .container;

  @media (min-width: 1440px) {
    padding: 0;
  }

  &--column {
    flex-direction: column;
  }

  &--row {
    flex-direction: row;
  }
}

.heading-font {
  font-family: $heading-font;
}