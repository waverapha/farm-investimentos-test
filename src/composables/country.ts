import { computed, reactive, ref } from "vue"
import type { Country } from "@/types/country"

const MAX_INITIAL_COUNTRIES_TO_DISPLAY = 5
const filter = ref('')
let allCountries = reactive(<Country[]>[])

export function useCountry() {
  const message = ref('')
  const countriesToDisplay = ref(MAX_INITIAL_COUNTRIES_TO_DISPLAY)
  const isCachingData = computed(() => message.value === 'Caching in progress')
  const filteredCountries = computed(() => {
    let countries = allCountries

    if (filter.value) {
      countries = allCountries.filter((country) => country.Country.toLowerCase().includes(filter.value.toLowerCase()))
    }

    return countries.slice(0, countriesToDisplay.value)
  })

  const fetchCountries = async () => {
    const response = await fetch('https://api.covid19api.com/summary')

    if (!response.ok) {
      throw new Error('Não foi possível obter os dados para a COVID-19 no momento. Atualize a página ou volte mais tarde.')
    }

    return await response.json()
  }

  const getCountries = async () => {
    try {
      const response: { Countries: Country[], Message: string, success?: boolean } = await fetchCountries()

      allCountries = response.Countries
      message.value = response.Message

      return response
    } catch (e: any) {
      message.value = e.message
    }
  }

  const loadMoreCountries = () => {
    countriesToDisplay.value += 5
  }

  return {
    MAX_INITIAL_COUNTRIES_TO_DISPLAY,
    filter,
    filteredCountries,
    isCachingData,
    message,
    getCountries,
    loadMoreCountries
  }
}