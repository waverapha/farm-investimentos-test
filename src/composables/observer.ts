export function useIntersectionObserver(handleIntersect: IntersectionObserverCallback, options?: IntersectionObserverInit) {
  const createObserver = () => {
    const observerOptions = Object.assign({
      root: null,
      rootMargin: "0px"
    }, options)

    return new IntersectionObserver(handleIntersect, observerOptions);
  }

  return {
    createObserver
  }
}