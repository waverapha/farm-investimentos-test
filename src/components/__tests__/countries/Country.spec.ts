import { describe, expect, it } from 'vitest'

import { screen } from '@testing-library/vue'

import { countryCardFactory, countryContainerFactory } from './helpers'
import { useCountry } from '@/composables/country'
import { countriesMock } from '../__mocks__/requests/countries-mock'

describe('country components tests', () => {
  it('expects MAX_INITIAL_COUNTRIES_TO_DISPLAY to be 5', () => {
    const { MAX_INITIAL_COUNTRIES_TO_DISPLAY } = useCountry()

    expect(MAX_INITIAL_COUNTRIES_TO_DISPLAY).toMatchInlineSnapshot('5')
  })

  it('should have same number of countries as MAX_INITIAL_COUNTRIES_TO_DISPLAY on start', async () => {
    countryContainerFactory()

    const countryToTest = 1
    const { MAX_INITIAL_COUNTRIES_TO_DISPLAY } = useCountry()

    const items = await screen.findAllByTestId('country-card')
    expect(items.length).toBe(MAX_INITIAL_COUNTRIES_TO_DISPLAY)

    const firstItemHeading = screen.getAllByRole('heading', { level: 3 })[countryToTest]
    expect(firstItemHeading.textContent).toBe(countriesMock.Countries[countryToTest].Country)
  }, { timeout: 1000 })

  it('should render card correctly', async () => {
    countryCardFactory()

    const countryName = countriesMock.Countries[0].Country

    const heading = screen.getByText(countryName)

    expect(heading.textContent).toBe(countryName)
  })
})