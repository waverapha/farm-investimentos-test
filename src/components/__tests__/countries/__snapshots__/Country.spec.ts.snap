// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`country components tests > expects MAX_INITIAL_COUNTRIES_TO_DISPLAY to be 5 1`] = `5`;
